### PR TITLE
Feature/reconciliation logic

### PR DESCRIPTION
**Pull request by FrazerBayley**
Implementing reconciliation logic inside SDK.
- Created asynchronous "subscribe" method
- "get" method is now blocking
- "init" method is no longer blocking

### CODE REVIEW

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocations.java [hunk 1]
```diff
@@ -49,4 +52,44 @@ JsonObject getGenomeFromAllocations() {
         genomeWrapped.add("genome", genome);
         return genomeWrapped;
     }
+
+    static JsonArray reconcileAllocations(JsonArray previousAllocations, @NotNull JsonArray currentAllocations) {
+        /*
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocations.java]: Can you use JavaDocs format in these?

https://google.github.io/styleguide/javaguide.html

Example:

/**
 * Returns an Image object that can then be painted on the screen. 
 * The url argument must specify an absolute {@link URL}. The name
 * argument is a specifier that is relative to the url argument. 
 * <p>
 * This method always returns immediately, whether or not the 
 * image exists. When this applet attempts to draw the image on
 * the screen, the data will be loaded. The graphics primitives 
 * that draw the image will incrementally paint on the screen. 
 *
 * @param  url  an absolute URL giving the base location of the image
 * @param  name the location of the image, relative to the url argument
 * @return      the image at the specified URL
 * @see         Image
 */
 public Image getImage(URL url, String name) {

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java [hunk 2]
```diff
@@ -2,26 +2,62 @@
...
 
+import ai.evolv.ascend.android.exceptions.AscendAllocationException;
+import okhttp3.Call;
+import okhttp3.Callback;
 import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+import timber.log.Timber;
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: Avoid direct dependency on platform specific libraries.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java [hunk 3]
```diff
@@ -2,26 +2,62 @@
...
+    private boolean contaminationSandbagged;
+    private AllocationStatus allocationStatus;
 
     Allocator(@NonNull AscendConfig config) {
-        this.client = config.getParticipantClient();
+        this.executionQueue = config.getExecutionQueue();
+        this.store = config.getAscendAllocationStore();
         this.config = config;
         this.ascendParticipant = config.getAscendParticipant();
+        this.confirmationSandbagged = false;
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: You can sent these in the declaration.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java [hunk 4]
```diff
@@ -2,26 +2,62 @@
 
 import android.support.annotation.NonNull;
 
-import ai.evolv.ascend.android.exceptions.AscendAllocationException;
-import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: Avoid third-party dependencies if possible.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java [hunk 5]
```diff
@@ -2,26 +2,62 @@
...
 import com.google.gson.JsonArray;
 import com.google.gson.JsonParser;
 
+import java.io.IOException;
+import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 
+import ai.evolv.ascend.android.exceptions.AscendAllocationException;
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: You probably want to move this to a generic namespace. (Sans android)

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java [hunk 6]
```diff
@@ -36,19 +72,112 @@ HttpUrl createAllocationsUrl() {
...
+                .build();
+
+        final SettableFuture<JsonArray> responseFuture = SettableFuture.create();
+
+        client.newCall(request).enqueue(new Callback() {
+            @Override public void onFailure(Call call, IOException e) {
+                resolveAllocationFailure(responseFuture);
+            }
+
+            @Override public void onResponse(Call call, Response response) {
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: I find putting annotations on their own line to be more readable. Totally subjective though.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java [hunk 7]
```diff
@@ -36,19 +72,112 @@ HttpUrl createAllocationsUrl() {
...
+                String body = "";
+                JsonArray allocations = new JsonArray();
+                try {
+                    try {
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: You're just catching the exception thrown in the outer try. You should move this try to the same level as the outer try.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java [hunk 8]
```diff
@@ -36,19 +72,112 @@ HttpUrl createAllocationsUrl() {
...
+                        ResponseBody responseBody = response.body();
+                        if (responseBody != null) {
+                            body = responseBody.string();
+                        }
+                        if (!response.isSuccessful()) {
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: Blank line after block.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java [hunk 9]
```diff
@@ -36,19 +72,112 @@ HttpUrl createAllocationsUrl() {
...
+                            throw new IOException("Unexpected response when making GET request: " + response + " using url: "
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: Use String.format

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java [hunk 10]
```diff
@@ -36,19 +72,112 @@ HttpUrl createAllocationsUrl() {
...
+                        }
+
+                        JsonParser parser = new JsonParser();
+                        allocations = parser.parse(body).getAsJsonArray();
+
+                        JsonArray previousAllocations = store.get();
+                        if (allocationsNotEmpty(previousAllocations)) {
+                            allocations = Allocations.reconcileAllocations(previousAllocations, allocations);
+                        }
+                    } catch(Exception e ) {
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: Extraneous whitespace.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java [hunk 11]
```diff
@@ -36,19 +72,112 @@ HttpUrl createAllocationsUrl() {
...
+                    }
+
+                    store.put(allocations);
+                    responseFuture.set(allocations);
+                    allocationStatus = AllocationStatus.RETRIEVED;
+
+                    if (confirmationSandbagged) {
+                        eventEmitter.confirm(allocations);
+                    }
+                    if (contaminationSandbagged) {
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: Blank line after block.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java [hunk 12]
```diff
@@ -36,19 +72,112 @@ HttpUrl createAllocationsUrl() {
...
+                        eventEmitter.contaminate(allocations);
+                    }
+
+                    // could throw an exception due to customer's action logic
+                    // always surface any customer implementation errors
+                    executionQueue.executeAllWithValuesFromAllocations(allocations);
+
+                } catch (AscendAllocationException e) {
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: Extraneous blank line

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java [hunk 13]
```diff
@@ -36,19 +72,112 @@ HttpUrl createAllocationsUrl() {
...
-            throw new AscendAllocationException(e.getMessage());
+            future.set(new JsonArray());
+
+            allocationStatus = AllocationStatus.FAILED;
+            executionQueue.executeAllWithValuesFromDefaults();
         }
+    }
 
-        return allocations;
+    static boolean allocationsNotEmpty(JsonArray allocations) {
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: You might want to rename this more generically.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/AscendAction.java [hunk 14]
```diff
@@ -0,0 +1,11 @@
+package ai.evolv.ascend.android;
+
+public interface AscendAction<T> {
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/AscendAction.java]: Should this not be something like ValueUpdateListener? What is "Action" in this context? Also, "Ascend" is redundant with the namespace.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/AscendAction.java [hunk 15]
```diff
@@ -0,0 +1,11 @@
+package ai.evolv.ascend.android;
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/AscendAction.java]: We should move the package to just ai.evolv, as this is hitting the participants.evolv.ai api.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/ExecutionQueue.java [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+import com.google.gson.JsonArray;
+
+import java.util.LinkedList;
+
+import ai.evolv.ascend.android.exceptions.AscendKeyError;
+import timber.log.Timber;
+
+class ExecutionQueue {
+
+    private final LinkedList<Execution> queue;
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/ExecutionQueue.java]: LinkedLists are not thread-safe. You should use a ConcurrentLinkedQueue or implement locking.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/HttpParticipantClient.java [hunk 17]
```diff
@@ -51,11 +47,11 @@
...
                     responseFuture.setException(e);
                 } finally {
-                    if (response != null) {
-                        response.close();
-                    }
+                    response.close();
                 }
             }
+
+
```
Comment (tfoster) [inline ascend-sdk/src/main/java/ai/evolv/ascend/android/HttpParticipantClient.java]: Extraneous whitespace.

File: example/build.gradle [hunk 18]
```diff
@@ -22,6 +29,7 @@ dependencies {
     implementation fileTree(dir: 'libs', include: ['*.jar'])
     implementation 'com.android.support:appcompat-v7:28.0.0'
     implementation 'com.android.support.constraint:constraint-layout:1.1.3'
+    implementation 'com.google.guava:guava:27.1-android'
```
Comment (tfoster) [inline example/build.gradle:32]: We should avoid dependency on third-party libs if possible.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocations.java (hunk 1 above)
Comment (FrazerBayley) [reply ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocations.java]: Ok, Im using proper javadoc in the reset of the project. Ill change this one though.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java (hunk 2 above)
Comment (FrazerBayley) [reply ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: Ill remove Timber and adopt SLF4J for this PR, and then implement pluggable http client in the next pr.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java (hunk 3 above)
Comment (FrazerBayley) [reply ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: 👍

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java (hunk 4 above)
Comment (FrazerBayley) [reply ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: Im going to keep the guava futures for now, ill see what I can do about removing Guava from the project.

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/ExecutionQueue.java (hunk 16 above)
Comment (FrazerBayley) [reply ascend-sdk/src/main/java/ai/evolv/ascend/android/ExecutionQueue.java]: Ok, ill adopt the ConcurrentLinkedQueue

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java (hunk 6 above)
Comment (FrazerBayley) [reply ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: Yah, lets be consistent

File: ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java (hunk 7 above)
Comment (FrazerBayley) [reply ascend-sdk/src/main/java/ai/evolv/ascend/android/Allocator.java]: The point of the nested try is to skip the logic outside of the inner try in the event the the inner try fails. I feel like this is the better way to go.

**Pull request merged**